kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/PulsarStorageProducerIdManagerImpl.java

```diff
@@ -0,0 +1,197 @@
+        return reader
+                .hasMessageAvailableAsync()
+                .thenCompose(hasMessageAvailable -> {
+                    if (hasMessageAvailable == null
+                            || !hasMessageAvailable) {
+                        return CompletableFuture.completedFuture(null);
+                    } else {
+                        CompletableFuture<Message<byte[]>> opMessage = reader.readNextAsync();
+                        return opMessage.thenCompose(msg -> {
+                            byte[] value = msg.getValue();
+                            long newId = new BigInteger(value).longValue();
+                            log.debug("{} Read {} from {}", this, newId, msg.getMessageId());
+                            nextId.set(newId);
+                            return readNextMessageIfAvailable(reader);
+                        });
+                    }
+                });
+    }
+
+    synchronized CompletableFuture<Void> ensureLatestData(boolean beforeWrite) {
+        if (currentReadHandle != null) {
+            if (beforeWrite) {
+                // we are inside a write loop, so
+                // we must ensure that we start to read now
+                // otherwise the write would use non up-to-date data
+                // so let's finish the current loop
+                log.debug("A read was already pending, starting a new one in order to ensure consistency");
+                return currentReadHandle
+                        .thenCompose(___ -> ensureLatestData(false));
+            }
+            // if there is an ongoing read operation then complete it
+            return currentReadHandle;
+        }
+        // please note that the read operation is async,
+        // and it is not execute inside this synchronized block
+        CompletableFuture<Reader<byte[]>> readerHandle = ensureReaderHandle();
+        final CompletableFuture<Void> newReadHandle =
+                readerHandle.thenCompose(this::readNextMessageIfAvailable);
+        currentReadHandle = newReadHandle;
+        return newReadHandle.whenComplete((a, b) -> {
+            endReadLoop(newReadHandle);
+            if (b != null) {
+                throw new CompletionException(b);
+            }
+        });
+    }
+
+    private synchronized void endReadLoop(CompletableFuture<?> handle) {
+        if (handle == currentReadHandle) {
+            currentReadHandle = null;
+        }
+    }
+
+    @Override
+    public synchronized CompletableFuture<Long> generateProducerId() {
+        Long booked = availableIdsLocally.poll();
+        if (booked != null) {
+            log.debug("Returning pre-allocated id {} for {}", booked, topic);
+            return CompletableFuture.completedFuture(booked);
+        }
+        log.debug("Allocating new block of ids for {}", topic);
+        // we could get rid of the Exclusive Producer if we had Message.getIndex()
+        // introduced in 2.9.0 https://github.com/apache/pulsar/pull/11553
```

Comment:
Just a note. I think we can add a separate PR to leverage `Message.getIndex` API for Pulsar 2.9.0 or higher in future.